text/0000-metric-label-set.md

```diff
@@ -0,0 +1,129 @@
+# Metric `LabelSet` specification
+
+**Status:** `proposed`
+
+Introduce a first-class `LabelSet` API type as a handle on a pre-defined set of labels.  
+
+## Motivation
+
+Labels are key-value pairs used across OpenTelemetry for categorizing spans (attributes, event fields) and metrics (labels).  Treatment of labels in the metrics API is especially important because of certain optimizations that take advantage of pre-defined labels.
+
+This optimization currently applies to metric instrument handles, the result of `GetHandle(Instrument, { Key : Value, ... })` on a metric instrument. By allowing the SDK to pre-compute information about the pair of `(Instrument, { Key : Value, ... })`, this reasoning goes, the metrics SDK has the opportunity make individual `Add()` and `Set()` operations on handles as fast as a few machine instructions.
+
+Adopting a first-class `LabelSet` extends the potential for optimization, significantly, for cases where a the `{ Key : Value, ... }` can be reused despite not re-using  `(Instrument, { Key : Value, ... })` pairs.
+
+The current API specifies one way to use set of labels for more than one measurement,
+
+```
+RecordBatch({ Key1: Value1,
+  	      Key2: Value2,
+	      ... },
+	    [ (Gauge, Value), 
+	      (Cumulative, Value),
+	      (Measure, Value),
+	      ... ])
+```
+
+This RFC proposes the new `LabelSet` concept is returned by an API named `Meter.DefineLabels({ Key: Value, ... })` which allows the SDK to potentially sort, canonicalize, or hash the set of labels once, allowing it to be re-used in several ways.  For example, the `RecordBatch` above can be written as one call to `DefineLabels` and inidividual operations.
+
+```
+Labels := meter.DefineLabels({ Key1: Value1,
+       	  		       Key2: Value2,
+			       ... })
+Cumulative.Add(Value, Labels)
```

Comment:
I think we said that we want always the operation to happen on the Handle not on the metric.

Reply:
This proposal is to support both handles and operations by `LabelSet`. There were several users, myself among them, who would prefer not to use handles, and label sets can be re-used in scenarios where handles maybe cannot, so it's still an important optimization. An example is where you want to compute a labelset once and then operate on a number of metrics.  There is no advantage to constructing handles if you'll use them once, but the label set can still be re-used here.

Then, if you've allowed that `RecordBatch` can record a number of (metric, labelset, value) tuples, it's just a syntactic convenience to be able to write metric.Record(labelset, value) to generate the equivalent tuple.